In [165]:
import pandas as pd
import numpy as np

In [166]:
raw_data_train = pd.read_csv("DataTrain_Preliminary.csv", sep = ";")

In [167]:
raw_data_train

duration protocol_type  service  flag src_bytes dst_bytes land  \
0             0           tcp  private    SH         0         0    0   
1             0           tcp  private    S0         0         0    0   
2             0           tcp     http    SF       285      3623    0   
3             0           tcp     http    SF       232       584    0   
4             1           tcp     smtp    SF      1080       327    0   
...         ...           ...      ...   ...       ...       ...  ...   
112441        0           tcp     http   REJ         0         0    0   
112442        0           tcp     http    SF       309       758    0   
112443        0           tcp     http    SF       363       721    0   
112444        0           tcp  discard  RSTO         0         0    0   
112445        0           tcp     http    SF       233      2538    0   

       wrong_fragment urgent hot  ... dst_host_srv_count  \
0                   0      0   0  ...                  1   
1                   0      0   0  ...                  5   
2                   0      0   0  ...                228   
3                   0      0   0  ...                255   
4                   0      0   0  ...                154   
...               ...    ...  ..  ...                ...   
112441              0      0   0  ...                255   
112442              0      0   0  ...                255   
112443              0      0   0  ...                255   
112444              0      0   0  ...                  7   
112445              0      0   0  ...                255   

       dst_host_same_srv_rate dst_host_diff_srv_rate  \
0                        0.01                   0.94   
1                        0.02                   0.08   
2                        1.00                   0.00   
3                        1.00                   0.00   
4                        0.58                   0.02   
...                       ...                    ...   
112441                   1.00                   0.00   
112442                   1.00                   0.00   
112443                   1.00                   0.00   
112444                   0.03                   0.06   
112445                   1.00                   0.00   

       dst_host_same_src_port_rate dst_host_srv_diff_host_rate  \
0                             0.95                           0   
1                             0.00                           0   
2                             0.01                           *   
3                             0.17                           *   
4                             0.00                           *   
...                            ...                         ...   
112441                        0.14                           *   
112442                        0.02                           *   
112443                        0.00                           0   
112444                        0.00                           0   
112445                        0.02                           *   

       dst_host_serror_rate dst_host_srv_serror_rate dst_host_rerror_rate  \
0                      0.95                        1                    0   
1                      1.00                        1                    0   
2                      0.00                        0                    0   
3                      0.00                        0                    0   
4                      0.00                        0                    0   
...                     ...                      ...                  ...   
112441                 0.00                        0                    *   
112442                 0.02                        *                    0   
112443                 0.00                        0                    0   
112444                 0.00                        0                    1   
112445                 0.00                        0                    0   

       dst_host_sr

In [168]:
raw_data_train.columns.values

array(['duration', 'protocol_type', 'service', 'flag', 'src_bytes',
       'dst_bytes', 'land', 'wrong_fragment', 'urgent', 'hot',
       'num_failed_logins', 'logged_in', 'num_compromised', 'root_shell',
       'su_attempted', 'num_root', 'num_file_creations', 'num_shells',
       'num_access_files', 'num_outbound_cmds', 'is_host_login',
       'is_guest_login', 'count', 'srv_count', 'serror_rate',
       'srv_serror_rate', 'rerror_rate', 'srv_rerror_rate',
       'same_srv_rate', 'diff_srv_rate', 'srv_diff_host_rate',
       'dst_host_count', 'dst_host_srv_count', 'dst_host_same_srv_rate',
       'dst_host_diff_srv_rate', 'dst_host_same_src_port_rate',
       'dst_host_srv_diff_host_rate', 'dst_host_serror_rate',
       'dst_host_srv_serror_rate', 'dst_host_rerror_rate',
       'dst_host_srv_rerror_rate', 'type_of_attack'], dtype=object)

In [169]:
row = raw_data_train.shape[0]
row

112446

In [170]:
column = raw_data_train.shape[1]
column

42

In [171]:
raw_data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112446 entries, 0 to 112445
Data columns (total 42 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   duration                     112446 non-null  object 
 1   protocol_type                111778 non-null  object 
 2   service                      111426 non-null  object 
 3   flag                         111593 non-null  object 
 4   src_bytes                    112446 non-null  object 
 5   dst_bytes                    112446 non-null  object 
 6   land                         112446 non-null  object 
 7   wrong_fragment               112446 non-null  object 
 8   urgent                       112446 non-null  object 
 9   hot                          112446 non-null  object 
 10  num_failed_logins            112446 non-null  object 
 11  logged_in                    112446 non-null  object 
 12  num_compromised              112446 non-null  object 
 13 

In [172]:
raw_data_train.isna().sum()

duration                          0
protocol_type                   668
service                        1020
flag                            853
src_bytes                         0
dst_bytes                         0
land                              0
wrong_fragment                    0
urgent                            0
hot                               0
num_failed_logins                 0
logged_in                         0
num_compromised                   0
root_shell                        0
su_attempted                      0
num_root                          0
num_file_creations                0
num_shells                        0
num_access_files                  0
num_outbound_cmds                 0
is_host_login                     0
is_guest_login                    0
count                             0
srv_count                         0
serror_rate                       0
srv_serror_rate                   0
rerror_rate                       0
srv_rerror_rate             

## We see that right now the dataset doesn't have a lot of missing values, so we will try to find some anomaly on the data

In [173]:
set(raw_data_train['type_of_attack'].values)

{'Denial of Service Attack',
 'ipsweep',
 'neptune',
 'nmap',
 'normal',
 'portsweep',
 'satan',
 'smurf'}

In [174]:
for col in raw_data_train.columns:
  print(col)
  print(raw_data_train[col].nunique())
  print(set(raw_data_train[col].values), "\n\n")

duration
2665
{'30297', '22516', '38849', '18158', '14362', '811', '6217', '12743', '8886', '31401', '236', '9555', '13488', '988', '26171', '15122', '40504', '7030', '22681', '7837', '1264', '10163', '3811', '1420', '3851', '6055', '3599', '372', '36653', '16170', '31905', '1723', '2711', '8784', '38342', '3405', '2714', '22901', '7979', '8624', '37532', '2705', '5322', '13919', '4848', '42', '3720', '2145', '8461', '5193', '2345', '1012', '2333', '3920', '18218', '4339', '3590', '1133', '4828', '17', '584', '5353', '15183', '41451', '42068', '13958', '9257', '282', '539', '730', '7783', '2455', '1605', '30357', '1061', '10995', '7800', '1014', '9411', '22396', '9065', '41782', '7780', '14680', '606', '31849', '5046', '1258', '31998', '6070', '30217', '3285', '3047', '10799', '72', '2684', '8380', '7405', '8179', '1100', '5310', '487', '103', '2976', '296', '21314', '25801', '6869', '385', '7379', '18661', '4764', '550', '41218', '30855', '20983', '235', '7510', '16513', '16845', '131

## Now we see that there is some anomaly in the data where the data contains * or 99999

We can print the column with anomaly

Columns with anomaly

In [175]:
columns_with_anomalies = []

In [176]:
for col in raw_data_train.columns:
    unique_values = set(raw_data_train[col].values)
    if "*" in unique_values or 99999 in unique_values:
        columns_with_anomalies.append(col)
print(columns_with_anomalies)

['duration', 'src_bytes', 'dst_bytes', 'land', 'wrong_fragment', 'urgent', 'hot', 'num_failed_logins', 'logged_in', 'num_compromised', 'root_shell', 'su_attempted', 'num_root', 'num_file_creations', 'num_shells', 'num_access_files', 'num_outbound_cmds', 'is_host_login', 'is_guest_login', 'count', 'srv_count', 'serror_rate', 'srv_serror_rate', 'rerror_rate', 'srv_rerror_rate', 'same_srv_rate', 'diff_srv_rate', 'srv_diff_host_rate', 'dst_host_count', 'dst_host_srv_count', 'dst_host_same_srv_rate', 'dst_host_diff_srv_rate', 'dst_host_same_src_port_rate', 'dst_host_srv_diff_host_rate', 'dst_host_serror_rate', 'dst_host_srv_serror_rate', 'dst_host_rerror_rate', 'dst_host_srv_rerror_rate']


Columns with NaN

In [177]:
columns_with_na = []

In [178]:
for col in raw_data_train.columns:
    if raw_data_train[col].isna().sum() > 0:
        columns_with_na.append(col)
print(columns_with_na)

['protocol_type', 'service', 'flag', 'dst_host_same_srv_rate', 'dst_host_diff_srv_rate', 'dst_host_same_src_port_rate', 'dst_host_serror_rate']


We should check that if there is any intersection of the columns with values of nan and columns with anomaly so that we can preprocess them further

In [179]:
for col in columns_with_na:
    if col in columns_with_anomalies:
        print(col)

dst_host_same_srv_rate
dst_host_diff_srv_rate
dst_host_same_src_port_rate
dst_host_serror_rate


We can see that there is intersection between the missing values and anomaly columns

For now we will check for the missing values

In [180]:
data_with_na = raw_data_train[columns_with_na]
data_with_na

protocol_type  service  flag  dst_host_same_srv_rate  \
0                tcp  private    SH                    0.01   
1                tcp  private    S0                    0.02   
2                tcp     http    SF                    1.00   
3                tcp     http    SF                    1.00   
4                tcp     smtp    SF                    0.58   
...              ...      ...   ...                     ...   
112441           tcp     http   REJ                    1.00   
112442           tcp     http    SF                    1.00   
112443           tcp     http    SF                    1.00   
112444           tcp  discard  RSTO                    0.03   
112445           tcp     http    SF                    1.00   

        dst_host_diff_srv_rate  dst_host_same_src_port_rate  \
0                         0.94                         0.95   
1                         0.08                         0.00   
2                         0.00                         0.01   
3                         0.00                         0.17   
4                         0.02                         0.00   
...                        ...                          ...   
112441                    0.00                         0.14   
112442                    0.00                         0.02   
112443                    0.00                         0.00   
112444                    0.06                         0.00   
112445                    0.00                         0.02   

        dst_host_serror_rate  
0                       0.95  
1                       1.00  
2                       0.00  
3                       0.00  
4                       0.00  
...                      ...  
112441                  0.00  
112442                  0.02  
112443                  0.00  
112444                  0.00  
112445                  0.00  

[112446 rows x 7 columns]

In [181]:
total_missing_values = row - data_with_na[data_with_na.isna().sum(1)==0].shape[0]
total_missing_values

3780

now we check the percentage of the missing values

In [182]:
print(round(total_missing_values/row*100, 2), "%")

3.36 %


we can see that the missing values is under 10%. This means we can drop the missing values since dropping it won't affect the data quality

In [183]:
data_no_na = raw_data_train.dropna()

In [184]:
row = data_no_na.shape[0]
row

108666

In [185]:
data_no_na

duration protocol_type  service  flag src_bytes dst_bytes land  \
0             0           tcp  private    SH         0         0    0   
1             0           tcp  private    S0         0         0    0   
2             0           tcp     http    SF       285      3623    0   
3             0           tcp     http    SF       232       584    0   
4             1           tcp     smtp    SF      1080       327    0   
...         ...           ...      ...   ...       ...       ...  ...   
112441        0           tcp     http   REJ         0         0    0   
112442        0           tcp     http    SF       309       758    0   
112443        0           tcp     http    SF       363       721    0   
112444        0           tcp  discard  RSTO         0         0    0   
112445        0           tcp     http    SF       233      2538    0   

       wrong_fragment urgent hot  ... dst_host_srv_count  \
0                   0      0   0  ...                  1   
1                   0      0   0  ...                  5   
2                   0      0   0  ...                228   
3                   0      0   0  ...                255   
4                   0      0   0  ...                154   
...               ...    ...  ..  ...                ...   
112441              0      0   0  ...                255   
112442              0      0   0  ...                255   
112443              0      0   0  ...                255   
112444              0      0   0  ...                  7   
112445              0      0   0  ...                255   

       dst_host_same_srv_rate dst_host_diff_srv_rate  \
0                        0.01                   0.94   
1                        0.02                   0.08   
2                        1.00                   0.00   
3                        1.00                   0.00   
4                        0.58                   0.02   
...                       ...                    ...   
112441                   1.00                   0.00   
112442                   1.00                   0.00   
112443                   1.00                   0.00   
112444                   0.03                   0.06   
112445                   1.00                   0.00   

       dst_host_same_src_port_rate dst_host_srv_diff_host_rate  \
0                             0.95                           0   
1                             0.00                           0   
2                             0.01                           *   
3                             0.17                           *   
4                             0.00                           *   
...                            ...                         ...   
112441                        0.14                           *   
112442                        0.02                           *   
112443                        0.00                           0   
112444                        0.00                           0   
112445                        0.02                           *   

       dst_host_serror_rate dst_host_srv_serror_rate dst_host_rerror_rate  \
0                      0.95                        1                    0   
1                      1.00                        1                    0   
2                      0.00                        0                    0   
3                      0.00                        0                    0   
4                      0.00                        0                    0   
...                     ...                      ...                  ...   
112441                 0.00                        0                    *   
112442                 0.02                        *                    0   
112443                 0.00                        0                    0   
112444                 0.00                        0                    1   
112445                 0.00                        0                    0   

       dst_host_sr

now we deal with the anomalies

In [186]:
data_with_anomalies = data_no_na[columns_with_anomalies]

In [187]:
data_with_anomalies.isna().sum()

duration                       0
src_bytes                      0
dst_bytes                      0
land                           0
wrong_fragment                 0
urgent                         0
hot                            0
num_failed_logins              0
logged_in                      0
num_compromised                0
root_shell                     0
su_attempted                   0
num_root                       0
num_file_creations             0
num_shells                     0
num_access_files               0
num_outbound_cmds              0
is_host_login                  0
is_guest_login                 0
count                          0
srv_count                      0
serror_rate                    0
srv_serror_rate                0
rerror_rate                    0
srv_rerror_rate                0
same_srv_rate                  0
diff_srv_rate                  0
srv_diff_host_rate             0
dst_host_count                 0
dst_host_srv_count             0
dst_host_s

We count the data with anomalies to decide wether will we drop it or impute it

In [189]:
anomalies_occurrences_per_row = ((raw_data_train == '*') | (raw_data_train == '99999') | (raw_data_train == 99999)).sum(axis=1)>0
total_anomalies = anomalies_occurrences_per_row.replace({True:1, False:0}).sum()
anomalies_percentage = round(total_anomalies/row*100, 2)
print(f'Anomalies Percentage on the given dataset after dropping na: {anomalies_percentage}%')

Anomalies Percentage on the given dataset after dropping na: 88.39%


As we can see that in the data we have 88% anomalies from the total data. This means we cannot simply drop the data we will need to impute the data